sensor/common/collectorruntimeconfig/handler.go

```diff
@@ -0,0 +1,125 @@
+	"github.com/stackrox/rox/pkg/logging"
+	"github.com/stackrox/rox/pkg/sync"
+	"github.com/stackrox/rox/sensor/common"
+	"github.com/stackrox/rox/sensor/common/message"
+)
+
+var (
+	log = logging.LoggerForModule()
+)
+
+// Store is a store for network graph external sources.
+//
+//go:generate mockgen-wrapper
+type Store interface {
+	CollectorConfigValueStream() concurrency.ReadOnlyValueStream[*sensor.CollectorConfig]
+}
+
+// Handler forwards the collector runtime config received from Central to Collectors.
+type Handler interface {
+	common.SensorComponent
+}
+
+type handlerImpl struct {
+	stopSig   concurrency.Signal
+	updateSig concurrency.Signal
+
+	collectorConfig            *sensor.CollectorConfig
+	collectorConfigProtoStream *concurrency.ValueStream[*sensor.CollectorConfig]
+
+	lock sync.Mutex
+}
+
+func (h *handlerImpl) Start() error {
+	go h.run()
+	return nil
+}
+
+func (h *handlerImpl) Stop(_ error) {
+	h.stopSig.Signal()
+}
+
+func (h *handlerImpl) Notify(common.SensorComponentEvent) {}
+
+func (h *handlerImpl) Capabilities() []centralsensor.SensorCapability {
+	// return []centralsensor.SensorCapability{centralsensor.NetworkGraphExternalSrcsCap}
+	return nil
+}
+
+func getCollectorConfig(msg *central.MsgToSensor) *sensor.CollectorConfig {
+	if clusterConfig := msg.GetClusterConfig(); clusterConfig != nil {
+		log.Info("Created fake collector config")
+		return &sensor.CollectorConfig{
+			NetworkConnectionConfig: &sensor.NetworkConnectionConfig{
+				EnableExternalIps: true,
+			},
+		}
+		// TODO: Once the collector config is a part of cluster config uncomment thi
+		// if config := clusterConfig.GetConfig(); config != nil {
+		//	if collectorConfig := config.GetCollectorConfig(); collectorConfig != nil {
+		//		return collectorConfig
+		//	}
+		//}
+	}
+
+	return nil
+}
+
+func (h *handlerImpl) ProcessMessage(msg *central.MsgToSensor) error {
+	log.Info("In ProcessMessage for collector runtime config")
+	collectorConfig := getCollectorConfig(msg)
+	if collectorConfig == nil {
+		return nil
+	}
+
+	log.Infof("Got collector config %+v", collectorConfig)
+	select {
+	case <-h.stopSig.Done():
+		return errors.New("could not process external network entities request")
+	default:
+		h.lock.Lock()
+		defer h.lock.Unlock()
+
+		h.collectorConfig = collectorConfig
+		h.updateSig.Signal()
+		return nil
+	}
+}
+
+func (h *handlerImpl) ResponsesC() <-chan *message.ExpiringMessage {
+	return nil
+}
+
+func (h *handlerImpl) run() {
+	for {
+		select {
+		case <-h.updateSig.Done():
+			h.pushConfigToValueStream()
+		case <-h.stopSig.Done():
+			return
+		}
+	}
+}
+
+func (h *handlerImpl) pushConfigToValueStream() {
+	h.lock.Lock()
+	defer h.lock.Unlock()
+
+	defer h.updateSig.Reset()
+
+	log.Info("In pushConfigToValueStream")
+	h.collectorConfigProtoStream.Push(h.collectorConfig)
+}
+
+func (h *handlerImpl) CollectorConfigValueStream() concurrency.ReadOnlyValueStream[*sensor.CollectorConfig] {
+	return h.collectorConfigProtoStream
```

Comment:
I am mostly just copying the pattern used to send the CIDR blocks to collector. I don't know if this is the best method.

Reply:
My use of the term CIDR block is a bit confusing here as it is used very little in the sensor code. That is perhaps a refactor that can be done later. The CIDR blocks are called `ip_networks` and are of type `IPNetworkList` in the `network_connection_iservice.proto` file. It has similar names in the go code.

The CIDR blocks are received by sensor from central in `sensor/common/externalsrcs/handler.go`  in the `ProcessMessage` function. There is a lot of processing of the data that comes in from central and that processed data is saved to two different maps. For collector config no processing is required except to extract it from the cluster config and it is simply saved as a collector config. Once the CIDR blocks have been processed they are pushed to `ipNetworkListProtoStream` which is of type `*concurrency.ValueStream[*sensor.IPNetworkList]`. This stream is read from in `sensor/common/networkflow/service/service_impl.go` and whenever new CIDR blocks appear in this stream or a new collector comes online the CIDR blocks are sent to collector. The actual function that does the sending is `sendExternalSrcsList`.

I think the advantage of using a handler that uses a stream such as this is that it handles concurrency issues and that the stream can potentially use multiple sources of information. The mechanism for sending the CIDR blocks is agnostic about the source of the CIDR blocks. While CIDR blocks are obtained from just one message type from central, collector configs could be obtained from cluster config messages from central or from helm charts. Using the stream would be helpful in this.

As for disadvantages, I think this approach is more complicated than my initial implementation, which had changes in fewer source code files. In that implementation there was no handler, only a service and the cluster config was received by the service and the service sent the collector config.

Reply:
This is a great explanation, thank you @JoukoVirtanen!